cmd/subctl/join.go

```diff
@@ -0,0 +1,238 @@
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package subctl
+
+import (
+	"errors"
+	"fmt"
+	"github.com/AlecAivazis/survey/v2"
+	"github.com/spf13/cobra"
+	submariner "github.com/submariner-io/submariner-operator/api/submariner/v1alpha1"
+	"github.com/submariner-io/submariner-operator/internal/cli"
+	"github.com/submariner-io/submariner-operator/internal/exit"
+	"github.com/submariner-io/submariner-operator/internal/restconfig"
+	"github.com/submariner-io/submariner-operator/pkg/broker"
+	"github.com/submariner-io/submariner-operator/pkg/join"
+	"github.com/submariner-io/submariner-operator/pkg/subctl/datafile"
+	"strings"
+)
+
+var joinFlags join.Options
+
+var joinCmd = &cobra.Command{
+	Use:     "join",
+	Short:   "Connect a cluster to an existing broker",
+	Args:    cobra.MaximumNArgs(1),
+	PreRunE: restConfigProducer.CheckVersionMismatch,
+	Run: func(cmd *cobra.Command, args []string) {
+		err := checkArgumentPassed(args)
+		exit.OnError(err)
+
+		subctlData, err := datafile.NewFromFile(args[0])
+		exit.WithMessage("Error loading the broker information from the given file", err)
+		fmt.Printf("* %s says broker is at: %s\n", args[0], subctlData.BrokerURL)
+
+		if joinFlags.ClusterID == "" {
+			joinFlags.ClusterID, err = askForClusterID()
+			exit.WithMessage("Error collecting information", err)
+		}
+
+		if joinFlags.ServiceCIDR == "" {
+			joinFlags.ServiceCIDR, err = askForCIDR("Service")
+			exit.WithMessage("Error detecting Service CIDR", err)
+		}
+
+		if joinFlags.ClusterCIDR == "" {
+			joinFlags.ClusterCIDR, err = askForCIDR("Cluster")
+			exit.WithMessage("Error detecting Cluster CIDR", err)
+		}
+
+		clientConfig, err := restConfigProducer.ClientConfig().ClientConfig()
+		_, clientset, err := restconfig.Clients(clientConfig)
+		exit.WithMessage("unable to set the Kubernetes cluster connection up", err)
+
+		gatewayNodesPresent, err := join.ListGatewayNodes(clientConfig)
+		if err != nil {
+			exit.WithMessage("Error getting gateway node", err)
+		}
+		var gatewayNode struct{Node string}
+		// If not Gateway nodes present, get all worker nodes and ask user to select one of them as gateway node
+		if !gatewayNodesPresent {
+			allWorkerNodeNames, err := join.GetAllWorkerNodeNames(clientset)
+			if err != nil {
+				exit.WithMessage("error listing worker nodes", err)
+			}
+			gatewayNode, err = askForGatewayNode(allWorkerNodeNames)
+			exit.WithMessage("Error getting gateway node", err)
+		}
+
+		brokerInfo, err := broker.ReadInfoFromFile(broker.InfoFileName)
+
+		err = join.SubmarinerCluster(*brokerInfo, joinFlags, restConfigProducer, cli.NewReporter(), gatewayNode)
+		exit.WithMessage("Error joining cluster", err)
```

Comment:
We should assume the status is complete and the returned error has been reported already.

```suggestion
		exit.OnError(err)
```